prompt2model/dataset_transformer/prompt_based.py

```diff
@@ -31,99 +34,72 @@ class PromptBasedDatasetTransformer(DatasetTransformer):
-            {"input_col": inputs, "output_col": outputs}
+        self.num_points_to_transform = num_points_to_transform
+        self.curr_failed_transforms = 0
+        self.max_allowed_failed_transforms = max_allowed_failed_transforms
+
+    def generate_task_explanation(self, prompt_spec: PromptSpec) -> str:
+        """Generate task explanation."""
+        task_explanation_prompt = construct_prompt_for_task_explanation(
+            prompt_spec.instruction, prompt_spec.examples
         )
-        return datasets.DatasetDict(dataset_dict)
+        return make_single_api_request(task_explanation_prompt, max_api_calls=10)
 
-    def transform_data(
-        self,
-        prompt_spec: PromptSpec,
-        dataset: datasets.Dataset,
-        num_points_to_transform: int,
-    ) -> datasets.DatasetDict:
-        """Transform the dataset according to the prompt_spec and dataset."""
+    def generate_plan(
+        self, task_explanation: str, dataset: datasets.Dataset, prompt_spec: PromptSpec
+    ) -> str:
+        """Generate plan for the task."""
         plan_prompt = self.plan_prompt_fn(
-            prompt_spec.instruction,
-            prompt_spec.examples,
-            dataset,
-            min(5, len(dataset)),
+            task_explanation, prompt_spec.examples, dataset
         )
-        self.plan = make_single_api_request(plan_prompt)
-
-        logger.info(f"Plan created. Plan: {self.plan}")
+        return make_single_api_request(plan_prompt, max_api_calls=10)
 
-        inputs = []
-        outputs = []
-
-        max_len = min(num_points_to_transform, len(dataset))
-        len_count = 0
+    def generate_transform_prompts(
+        self,
+        task_explanation: str,
+        dataset: datasets.Dataset,
+        prompt_spec: PromptSpec,
+    ) -> list[str]:
+        """Get transform prompts for each row in the dataset."""
         transform_prompts = []
-        for row in dataset:
+        for i in range(min(self.num_points_to_transform, len(dataset))):
+            row = dataset[i]
             transform_prompt = self.transform_prompt_fn(
-                prompt_spec.instruction,
-                prompt_spec.examples,
-                self.plan,
-                row,
+                task_explanation, row, self.plan, prompt_spec.examples
             )
             transform_prompts.append(transform_prompt)
+        return transform_prompts
 
-            len_count += 1
-            if len_count >= max_len:
-                break
+    def generate_responses(self, transform_prompts_batch: list[str]) -> list[str]:
+        """Generate responses for the transform prompts."""
 
-        async def generate_responses(transform_prompts):
-            responses = await api_tools.default_api_agent.generate_batch_completion(
+        async def generate_responses_async(transform_prompts):
+            """Generate responses asynchronously using the specified model."""
+            responses = await api_tools.APIAgent(
```

Comment:
can we expose APIAgent as a function parameter with this APIAgent being the default choice? This is currently too restrictive.

Reply:
@viswavi let me know if the _generate_responses()_  function should have just model_name passed in or is it better to  pass in the entire api_agent?